src/Distribution/Nixpkgs/Fetch.hs

```diff
@@ -72,16 +72,18 @@ fromDerivationSource :: DerivationSource -> Source
 fromDerivationSource DerivationSource{..} = Source derivUrl derivRevision (Certain derivHash) "."
 
 -- | Fetch a source, trying any of the various nix-prefetch-* scripts.
-fetch :: forall a. (String -> MaybeT IO a)      -- ^ This function is passed the output path name as an argument.
+fetch :: forall a.
+         Bool                                   -- ^ If True, fetch submodules when the source is a git repository
+      -> (String -> MaybeT IO a)                -- ^ This function is passed the output path name as an argument.
                                                 -- It should return 'Nothing' if the file doesn't match the expected format.
                                                 -- This is required, because we cannot always check if a download succeeded otherwise.
       -> Source                                 -- ^ The source to fetch from.
       -> IO (Maybe (DerivationSource, a))       -- ^ The derivation source and the result of the processing function. Returns Nothing if the download failed.
-fetch f = runMaybeT . fetchers where
+fetch s f = runMaybeT . fetchers where
   fetchers :: Source -> MaybeT IO (DerivationSource, a)
   fetchers source = msum . (fetchLocal source :) $ map (\fetcher -> fetchWith fetcher source >>= process)
     [ (False, "url", [])
-    , (True, "git", ["--fetch-submodules"])
+    , (True, "git", if s then ["--fetch-submodules"] else [])
```

Comment:
IMHO if you name `s` a bit more descriptively like `optSubmodules` which is also consistent with the call site then the hlint suggestion of using: 
```haskell
[ "--fetch-submodules" | optSubmodules ]
``` 
is quite clear and understandable. It's a bit like the `optional` function in Nix.

Reply:
That's a good suggestion, thank you.